### ISSUE TITLE
"ipfs add" calculate CID and check for it *before uploading* to API server.

### ISSUE THREAD

**Issue by kallisti5** [ipfs/go-ipfs#7586]
ipfs add should calculate the CID and check for it's existence on the api server / network *before* copying the data over.

An example scenario:

  * I rsync 67GiB worth of files to my system
  * I ```ipfs --api=$IPFS_API_SERVER add -r /cache/*```  to make them available.

In the above scenario, my remote ipfs client seemingly needs to upload 67.88GiB of files to the API server every time I sync to IPFS.  Ideally the local ipfs binary could quickly generate a CID and check for the CID's existance *before* transferring the data to the remote API server.  Even in local deployments this could save time since IPFS unnecessarily transferring files.

![ipfs](https://user-images.githubusercontent.com/100110/89739346-8956ff00-da45-11ea-89bf-1b92bf7c84de.png)

**Comment by aschmahmann** [comment]
@kallisti5 this is mostly a function of the "client" being very thin and would likely be a bit of a pain (although not infeasible) to implement without breaking changes. A few notes:

1) In order to add a file/directory to IPFS it needs to go through the the process of being converted into IPLD blocks (i.e. chunking + UnixFSifying the data).
    - This means that in order to calculate the CID we'd need to do all the chunking and processing work on the client end, which might be costly
    - In order to not duplicate chunking work we'd need to use an API endpoint that uses blocks instead of files, which is a breaking change from how `ipfs add` works today
2) IMO this is going about things the wrong way. Instead of trying to make the client "smarter" instead data should just be imported to IPFS as locally as possible in order to take advantage of IPFS's properties (e.g. data deduplication, downloading data from many parties, download resumption, etc.)
   - For example, if you're trying to import 67GiB of data to a go-ipfs instance that's on the "Server" why `ipfs add --api` from the "Client" instead of either:
       - Adding the data to the server and then adding it to go-ipfs
       - Running a second IPFS instance on the "Client" and after adding files on the Client running an `ipfs add` on the server (performance of the transfer can be boosted by having the Client do an `ipfs swarm connect /multiaddress/of/server` beforehand.)

> Even in local deployments this could save time since IPFS unnecessarily transferring files.

I agree that there is over-coupling of the local CLI interactions with the remote CLI + HTTP API. If we could separate them out more cleanly we could take advantage of their differences. For example, when working with a local daemon we could send file names instead of byte streams to the daemon so that we don't even have to read the data on the client end. However, for this to be worthwhile we'd have to get some benchmarks showing that streaming the data over HTTP from one process in the same machine to another makes a measurable impact on `ipfs add` performance.

**Comment by kallisti5** [comment]
> For example, if you're trying to import 67GiB of data to a go-ipfs instance that's on the "Server" why ipfs add --api from the
> "Client" instead of either:
>
>    Adding the data to the server and then adding it to go-ipfs
>    Running a second IPFS instance on the "Client" and after adding files on the Client running an ipfs add on the server >>  (performance of the transfer can be boosted by having the Client do an ipfs swarm connect /multiaddress/of/server beforehand.)

Mostly because most folks architect around microservices these days :-)    Here are a few use cases:

  * **Home usage**: I have a Synology NAS with docker, running the IPFS daemon there... exposing it via NAT forwarding, and using it as a gateway for nodes on my network (as well as my phone). To add something from any machine on my network, I add --api to the client's ipfs settings and can publish things to the IPFS network.
  * **Server usage**: Deploying IPFS to a container in kubernetes, having multiple other containers use the IPFS daemon (or cluster) as a "gateway" to ipfs services.


As for alternatives, Is there any way to generate the CID multihash outside of the ipfs daemon for individual files, check for it's existence on the IPFS node, then only adding the file when missing? I see a few multihash projects in rust 

Not 100% sure how adding a folder of files would work if I can't ```ipfs add -r directory/*``` though...

Someone mentioned their rsync-to-ipfs script here which gives some clues how to generate folder CID's external to the IFPS daemon, but it's pretty complicated to just sync some files to IPFS :-|

https://github.com/RubenKelevra/rsync2ipfs-cluster/blob/master/bin/rsync2cluster.sh

```shell
#get new rootfolder CIDs
ipfs_mfs_folder_cid=$(ipfs_api files stat --hash "/$ipfs_folder") || fail 'repo folder (IPFS) CID could not be determined after update is completed' 400

echo -ne ":: publishing new root-cid to DHT..."
ipfs_api dht provide --timeout 3m "$ipfs_mfs_folder_cid" > /dev/null || warn 'Repo folder (IPFS) could not be published to dht after update\n' -n
echo "done."
```

**Comment by kallisti5** [comment]
tldr; this would all be a lot simpler if ```ipfs add -r directory/*``` would identify CID's as "already existing" on the target IPFS node and skip uploading them.    You would be able to keep massive directories in-sync to IPFS with a few lines of code vs 600+ line scripts manually managing data.

**Comment by lidel** [comment]
@kallisti5 I only glanced at your use case, but perhaps all you need is:

- `ipfs add --offline --only-hash <file>` run against local repo will return CID without sending the data over the net or adding it to local datastore
- `ipfs block get --offline <cid>`  (at remote node) will return error if CID is not already present in the datastore

**Comment by kallisti5** [comment]
Oh... ```ipfs block stat``` does indeed seem pretty close to a solution.
Is there a way to build a directory CID without uploading all of the files?

aka.. ```ipfs add -r mydir``` builds a directory of CID's a filenames and returns a CID. The only way i'm aware of to build it is to do an ```ipfs add -r ...``` (which will upload everything)

**Comment by lidel** [comment]
> Is there a way to build a directory CID without uploading all of the files?

Yes, you can use `ipfs object patch` and `add-link` to build any unixfs directory tree you want, all without fetching the actual content of CIDs you use while building it:

- https://discuss.ipfs.io/t/adding-a-file-while-preserving-filename-using-js-ipfs/2060/2?u=lidel
- https://discuss.ipfs.io/t/how-can-attach-cid-to-new-node/7534/3?u=lidel

**Comment by kallisti5** [comment]
This definitely seems like a potential solution, but it's still pretty slow.

```shell
NEWFOLDER=""
find /cache -type f -not -name ".*" | while read line; do
        HASH=$(ipfs add -q --offline --only-hash $line)
        BARENAME=$(echo $line | sed -e 's/^\/cache\///')
        NEWFOLDER=$(printf "$NEWFOLDER\n$HASH:$BARENAME")
        ipfs --api=$IPFS_API_SERVER block stat --offline $HASH > /dev/null 2>&1
        if [[ $? -eq 0 ]]; then
                echo "[HAVE] $HASH $BARENAME"
        else
                echo "[NEED] $HASH $BARENAME"
                ipfs --api=$IPFS_API_SERVER add $line
        fi
done
```

Now you take NEWFOLDER which has:
```CID:relative/path/to/file```

Then you patch an empty folder, appending each resulting CID on top of the last?
(generating a *TON* of unused folder structures on the IPFS node.

I think this example helps to demonstrate how quickly a basic task of keeping a large number of moderately sized files can quickly overwhelm the CLI tools.

If ipfs add -r  locally calculated and checked for the remote file before sending it over the wire, the "remote IPFS node" use case gets a lot less painful.

I get that the core demographic is desktops running IPFS, but in any kind of server or "central IPFS node" environment things get painful quickly. (I know of IPFS cluster, but it would suffer the same issues in theory unless you're calculating and uploading directly from a cluster member)

**Comment by kamaradclimber** [comment]
Would also be interested in such solution for the following use case: making backup of data stored by containers.
The idea is to make a frequent backup of data stored by a given container while avoiding to resend data that has not changed (we can assume that most files won't change between two backups).

Bonus point if the solution works with ipfs-cluster :)

**Comment by RubenKelevra** [comment]
@kamaradclimber well, this sounds more like a use case for a local ipfs node that can access the containers via the filestore option. You can create the CID and add it to the cluster, while the filestore will provide the data when the cluster requests them.

**Comment by kamaradclimber** [comment]
Thanks for your answer. I think I've not explained correctly the use case. I'm not trying to distribute the container image but rather to make frequent backups of some of the data an app instance (which happens to run in a container) is producing.

**Comment by kallisti5** [comment]
A quick addition on "why".

IPFS seems like it would be a *powerful* tool to replicate a software package repository (think apt repo).  It has:

* Built in transparent signature checking
* Ability for users to mirror packages by pinning an IPNS entry
* Deduplication to reduce space utilization
* Any user can "setup" a mirror via an IPFS gateway
* Any gateway could be used as a mirror, accessing content pinned by all users

This use case is intense (and seemingly way more useful than the current hosting a "static website" design plans).  Who needs rsync mirrors when you can just have folks pin content on IPFS?

**Comment by lidel** [comment]
Back to the original feature request:

> "ipfs add" calculate CID and check for it *before* uploading to API server

go-ipfs now supports `ipfs dag export|import` commands, which make it possible to implement  requested behavior in userland:

1. Run `ipfs add` on machine (A), it  will produce CID. If you don't want to duplicate data, enable [filestore](https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#ipfs-filestore) and do `ipfs add --nocopy`
2. Run something like  `ipfs files stat /ipfs/{cid} --with-local` on machine (B) to see if it is  100% present on (B)
3. If not, `ipfs dag export` on (A), copy CAR archive to (B)  and  do `ipfs dag import` there.

You can do this today with go-ipfs 0.9.0

We could introduce `ipfs add --create-car` which creates a CAR archive first, and then uploads it to API via `/api/v0/dag/import` (+ leaves CAR artifact as side-effect), but this seems like a lot of work to account for a niche use case, which could be solved in way more flexible way in userland. 

Perhaps a better home for this feature is a tool/script built on top of  `ipfs` command?

**Comment by liqsliu** [comment]
my solution: check `curl -m 2 --max-filesize 1 -s -o /dev/null -w '%{http_code}'  https://bafkreibuuyw5r6ji4u4ysi3w6f45mfbvyravqb77vtp5kootzncpnfqpbm.ipfs.infura-ipfs.io/?filename=file_781.webp`
